go/nbs/table_persister.go

```diff
@@ -55,6 +58,89 @@ func (csbc chunkSourcesByAscendingCount) Less(i, j int) bool {
 }
 func (csbc chunkSourcesByAscendingCount) Swap(i, j int) { csbc[i], csbc[j] = csbc[j], csbc[i] }
 
+type compactionPlan struct {
+	sources             chunkSources
+	chunkDataLens       []uint64
+	mergedIndex         []byte
+	chunkCount          uint32
+	totalCompressedData uint64
+}
+
+func (cp compactionPlan) lengths() []byte {
+	lengthsStart := uint64(cp.chunkCount) * prefixTupleSize
+	return cp.mergedIndex[lengthsStart : lengthsStart+uint64(cp.chunkCount)*lengthSize]
+}
+
+func (cp compactionPlan) suffixes() []byte {
+	suffixesStart := uint64(cp.chunkCount) * (prefixTupleSize + lengthSize)
+	return cp.mergedIndex[suffixesStart : suffixesStart+uint64(cp.chunkCount)*addrSuffixSize]
+}
+
+func planCompaction(sources chunkSources) (plan compactionPlan) {
+	var totalUncompressedData uint64
+	for _, src := range sources {
+		plan.chunkCount += src.count()
+		totalUncompressedData += src.uncompressedLen()
+	}
+	plan.sources = sources
+
+	lengthsPos := uint64(plan.chunkCount) * prefixTupleSize
+	suffixesPos := lengthsPos + uint64(plan.chunkCount)*lengthSize
+	plan.mergedIndex = make([]byte, indexSize(plan.chunkCount)+footerSize)
+	plan.chunkDataLens = make([]uint64, len(sources))
+
+	prefixIndexRecs := make(prefixIndexSlice, 0, plan.chunkCount)
+	var ordinalOffset uint32
+	for i, src := range sources {
+		idx := src.index()
+
+		// Add all the prefix tuples from this index to the list of all prefixIndexRecs, modifying the ordinals such that all entries from the 1st item in sources come after those in the 0th and so on.
+		for j, prefix := range idx.prefixes {
+			rec := prefixIndexRec{prefix: prefix, order: ordinalOffset + idx.ordinals[j]}
+			prefixIndexRecs = append(prefixIndexRecs, rec)
+		}
+		ordinalOffset += src.count()
+
+		// Calculate the amount of chunk data in |src|
+		plan.chunkDataLens[i] = idx.offsets[src.count()-1] + uint64(idx.lengths[src.count()-1])
```

Comment:
the size is the just the first byte of the index. let's we calculate that from chunkCount when we parse an index. lets use that.